File: numerical-analysis/equation/python/bisection.py
import math

N, TOL, a, b = map(float, input().split())
N = int(N)

def f(x: float) -> float:
    return x * math.exp(-x) + math.pow(x, 3.0) + 1

k, FA = 1, f(a)
while (k <= N):
    c = a + (b - a) / 2
    FC = f(c)
    if (FC == 0 or (b - a) / 2 < TOL):
        print(f"{c}")
        exit(0)
    if (FA * FC < 0):
        b = c
    else:
        a, FA = c, FC # [WARNING] is FA need to be updated? 
    k = k + 1

print(f"Method failed after {N} iterations")
exit(-1)

File: numerical-analysis/equation/c++/bisection.cpp
#include <cmath>
#include <iostream>

double f(double x) {
    return x * exp(-x) + pow(x, 3) + 1;
}

int main() {
    int N;
    double TOL, a, b;
    std::cin >> N >> TOL >> a >> b;
    
    int k = 1;
    double FA = f(a);

    while (k <= N) {
        double c = a + (b - a) / 2;
        double FC = f(c);
        if (FC == 0 || (b - a) / 2 < TOL) {
            std::cout << c << std::endl;
            return 0;
        }
        if (FA * FC < 0) {
            b = c;
        } else {
            a = c;
            FA = FC;
        }
        k = k + 1;
    }

    std::cout << "Method failed after " << N << " iterations" << std::endl;

    return 0;
}